### PR TITLE
Define Firefox versions for several css files (2nd pass)

### PR DESCRIPTION
**Pull request by vinyldarkscratch**
This is to help the efforts of #3126 and extend upon #3439 by defining Firefox versions for many CSS entries.  Looking through the Git mirror of `mozilla-central` (`gecko-dev`), I've found the following commits where the CSS rules originally appear.

# Found and Versioned

## Exact Commits of Implementation

These are the earliest commits that add support for the feature.

**css.properties.align-content.flex_context.start_end, css.properties.align-content.flex_context.baseline, css.properties.align-self.flex_context.start_end, css.properties.align-self.flex_context.baseline, css.properties.justify-content.flex_context.start_end, css.properties.justify-content.flex_context.baseline**

```
commit cb1d1d16b19e4f6a1c3785fd4195a4dd9b3f4788
https://hg.mozilla.org/mozilla-central/rev/f7d6e92ba3df
Author: Mats Palmgren <mats@mozilla.com>
Date:   Tue Nov 3 15:18:05 2015 +0100
Ver.:   45.0a1
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=1176782
```

**css.properties.contain**

```
commit 95a87c9836c143ba4c155d9cf6a66ece7662981e
https://hg.mozilla.org/mozilla-central/rev/8422e49e2ab3
Author: Kyle Zentner <kzentner@mozilla.com>
Date:   Thu Jun 4 16:38:00 2015 +0200
Ver.:   41.0a1
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=1170173
```

**css.selectors.-moz-color-swatch**

```
commit 574b47d0902fad8fc02a97fada8285f0337703aa
https://hg.mozilla.org/mozilla-central/rev/3975fe261080
Author: Arnaud Bienner <arnaud.bienner@gmail.com>
Date:   Wed Jul 10 00:25:27 2013 +0200
Ver.:   27.0a1
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=875275
```

**css.selectors.indeterminate**

```
commit 8ddfeb2aa297b5d7f8e516d82a01cf626637fae9
https://hg.mozilla.org/mozilla-central/rev/58c758340aa5
Author: Michael Ventnor <ventron.bugzilla@gmail.com>
Date:   Wed Feb 11 14:19:37 2009 +1300
Ver.:   2
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=477531
```

**css.selectors.-moz-range-progress**

```
commit ce9ffb26288dd5a69e97b5c0796bb2bc14d485f7
https://hg.mozilla.org/mozilla-central/rev/805aff291ec2
Author: Jonathan Watt <jwatt@jwatt.org>
Date:   Tue Mar 26 23:04:41 2013 +0000
Ver.:   22.0a1
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=854538
```

**css.selectors.-moz-range-thumb**

```
commit 54936aa0ff098ab3902631e2351bd90c6ae3e8e8
https://hg.mozilla.org/mozilla-central/rev/0f9734e836ec
Author: Wes Johnston <wjohnston@mozilla.com>
Date:   Sat Feb 16 17:57:21 2013 +0000
Ver.:   21.0a1
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=838256
```

**css.selectors.-moz-range-track**

```
commit fd7314abae79ece3cdce33cfe4651d35c6a22f9a
https://hg.mozilla.org/mozilla-central/rev/7db39954cc10
Author: Jonathan Watt <jwatt@jwatt.org>
Date:   Sat Feb 16 17:57:21 2013 +0000
Ver.:   21.0a1
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=838256
```

## Evidence of Use in 1.0

These commits show the properties being used in Phoenix/Firebird (Firefox 0.x) or Firefox 1.0, or date far back enough.

**css.properties.-moz-user-input, css.properties.-moz-user-input-none, css.properties.-moz-user-input.enabled, css.properties.-moz-user-input.disabled, css.properties.-moz-user-input.auto**

```
commit cfceba7dca7e62634614e1daae01d0875ae1b70a
Author: dbaron%fas.harvard.edu <dbaron%fas.harvard.edu>
Date:   Wed Sep 6 02:17:26 2000 +0000

commit 492bf2f145d78d93c2e5019838ed92e3e51c2f8b
Author: peterl%netscape.com <peterl%netscape.com>
Date:   Sun Jul 18 00:14:08 1999 +0000
```

**css.properties.caption-side.non_standard_values**

```
commit 00c5637f3b356177485f6e63d10d38c373fb87ff
Author: janc%netscape.com <janc%netscape.com>
Date:   Tue Oct 27 02:05:42 1998 +0000
```

**css.properties.user-modify**

```
commit a8e65ff2a6c79d67aa71bd31c04b6d6c0477daab
Author: peterl%netscape.com <peterl%netscape.com>
Date:   Fri Sep 3 23:40:35 1999 +0000
```

**css.properties.user-select.auto**

```
commit 096c23050e22a181b8eed4426cbd3c2bbd31b7e0
Author: pierre%netscape.com <pierre%netscape.com>
Date:   Tue Nov 23 01:07:56 1999 +0000
```

**css.properties.user-select, css.properties.user-select.all, css.properties.user-select.none, css.properties.user-select.text**

```
commit 2507785091be9c455507806b25cc515c934f3648
Author: peterl%netscape.com <peterl%netscape.com>
Date:   Fri Sep 3 23:33:06 1999 +0000
```

**css.properties.text-rendering, css.properties.text-rendering.auto, css.properties.text-rendering.geometricPrecision**

```
commit 5c676d4a472c56281b102fdef1336d42f04fd466
Author: alex.fritze%crocodile-clips.com <alex.fritze%crocodile-clips.com>
Date:   Sat Feb 7 12:39:26 2004 +0000
```

**css.properties.visibility, css.properties.visibility.collapse**

```
commit 18bb385451042e8ae79b15e143f3cec04b161670
Author: peterl%netscape.com <peterl%netscape.com>
Date:   Sun Jul 18 00:32:32 1999 +0000

Includes mention of "collapse" in this commit
```

**css.selectors.-moz-page, css.selectors.-moz-page-sequence**

```
commit f6b1702067299e0f985a02f3b3bc939c7f336340
Author: dbaron%fas.harvard.edu <dbaron%fas.harvard.edu>
Date:   Sun Nov 17 15:37:56 2002 +0000
```

**css.properties.-moz-binding**

```
commit 8c105459aec5d640590895c03975e4bbe0af0a41
Author: dbaron%fas.harvard.edu <dbaron%fas.harvard.edu>
Date:   Wed Sep 6 02:15:18 2000 +0000
```

**css.selectors.-moz-only-whitespace**

```
commit 6de447be6a438f24d5209e7296908876a200cd53
Author: bzbarsky%mit.edu <bzbarsky%mit.edu>
Date:   Fri Dec 3 04:25:21 2004 +0000
```

## Estimated by Date

These commits reveal a feature's implementation, but for some reason cannot be tracked down in Mercurial and do not have any milestone data in Bugzilla.  Their versions have been estimated based upon their date.

**css.properties.mask**

```
commit 5cd5d599181a6f2e3d7cd4ad791775c949aff1f7
Author: tor%cs.brown.edu <tor%cs.brown.edu>
Date:   Fri Jan 20 17:00:43 2006 +0000
Ver.?:  2
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=316764
```

**css.selectors.-moz-scrolled-page-sequence**

```
commit 301b41101358f45730cdfe4d69e9423770deedf5
Author: roc+%cs.cmu.edu <roc+%cs.cmu.edu>
Date:   Thu Apr 28 22:14:16 2005 +0000
Ver.?:  1.5
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=290752
```

**css.selectors.read-only, css.selectors.read-write**

```
commit 573c3b6d32908eb166b2bd08b91b26bd7832fc4b
Author: doronr%us.ibm.com <doronr%us.ibm.com>
Date:   Tue Oct 18 21:16:27 2005 +0000
Ver?.:  1.5
Bug:    https://bugzilla.mozilla.org/show_bug.cgi?id=302188
```

# Skipped

## Already Added

- css.properties.cursor.url_positioning_syntax

## Not in Compat Table

- css.properties.background-image.image-rect
- css.properties.background-image.any_image
- css.properties.place-content.flex_context
- css.selectors.blank
- css.types.image.gradient.linear-gradient

## Unsure of Meaning

(AKA This is a shorthand for multiple properties or an accepted value that I don't understand, can someone explain what this means?)

- css.properties.border-radius.4_values_for_4_corners

# In Progress

# To Do

- css.properties.align-items.grid_context.start_end
- css.properties.background-image.element
- css.properties.background-image.image-rect
- css.properties.outline-color.invert
- css.properties.place-content.flex_context
- css.properties.transition-property.IDENT_value
- css.selectors.placeholder-shown.non_text_types
- css.selectors.read-write.matches_editable_elements
- css.types.image
- css.types.image.gradient
- css.types.length.ex

# Information

Servo: Initial Add

```
commit cce5500148d21ed7e9e813b2faa2aecb40867cb6
https://hg.mozilla.org/mozilla-central/rev/55e5cc97e3ea
Author: Dave Herman <dherman@mozilla.com>
Date:   Wed Feb 8 11:29:12 2012 -0800

push 31307
Changeset: https://hg.mozilla.org/mozilla-central/pushloghtml?changeset=c3a5180a727e
Commit:    https://hg.mozilla.org/mozilla-central/rev/94079d43835f
Date:      Sat, 04 Feb 2017 00:59:06 +0000
Version:   54.0a1
```

### CODE REVIEW

**Comment by vinyldarkscratch** [comment]
This one has proven more fun than I originally thought -- I'm happy to be working on this one!

File: css/properties/border-bottom-left-radius.json [hunk 1]
```diff
@@ -58,10 +58,32 @@
...
+                    "type": "preference",
+                    "name": "layout.css.prefixes.webkit",
+                    "value_to_set": "true"
+                  }
+                ]
+              },
+              {
+                "alternative_name": "-moz-border-radius-bottomleft",
+                "version_added": "4",
+                "version_removed": "12"
```
Comment (ExE-Boss) [inline css/properties/border-bottom-left-radius.json]:
```suggestion
                "version_removed": "14"
```

File: css/properties/border-bottom-right-radius.json [hunk 2]
```diff
@@ -58,10 +58,32 @@
...
+                    "type": "preference",
+                    "name": "layout.css.prefixes.webkit",
+                    "value_to_set": "true"
+                  }
+                ]
+              },
+              {
+                "alternative_name": "-moz-border-radius-bottomright",
+                "version_added": "4",
+                "version_removed": "12"
```
Comment (ExE-Boss) [inline css/properties/border-bottom-right-radius.json]:
```suggestion
                "version_removed": "14"
```

File: css/properties/border-radius.json [hunk 3]
```diff
@@ -50,10 +50,17 @@
...
-            },
+            "firefox_android": [
+              {
+                "version_added": "4",
+                "notes": "Prior to Firefox 50, border styles of rounded corners (with <code>border-radius</code>) were always rendered as if <code>border-style</code> was <code>solid</code>. This has been fixed in Firefox 50."
+              },
+              {
+                "prefix": "-moz-",
+                "version_added": "4",
+                "version_removed": "12"
```
Comment (ExE-Boss) [inline css/properties/border-radius.json]:
```suggestion
                "version_removed": "14"
```

File: css/properties/border-top-left-radius.json [hunk 4]
```diff
@@ -58,10 +58,32 @@
...
+                    "type": "preference",
+                    "name": "layout.css.prefixes.webkit",
+                    "value_to_set": "true"
+                  }
+                ]
+              },
+              {
+                "alternative_name": "-moz-border-radius-topleft",
+                "version_added": "4",
+                "version_removed": "12"
```
Comment (ExE-Boss) [inline css/properties/border-top-left-radius.json]:
```suggestion
                "version_removed": "14"
```

File: css/properties/border-top-right-radius.json [hunk 5]
```diff
@@ -58,10 +58,32 @@
...
+                    "type": "preference",
+                    "name": "layout.css.prefixes.webkit",
+                    "value_to_set": "true"
+                  }
+                ]
+              },
+              {
+                "alternative_name": "-moz-border-radius-topright",
+                "version_added": "4",
+                "version_removed": "12"
```
Comment (ExE-Boss) [inline css/properties/border-top-right-radius.json]:
```suggestion
                "version_removed": "14"
```

**Comment by vinyldarkscratch** [comment]
Review?  @ddbeck

File: css/properties/background-image.json [hunk 6]
```diff
@@ -22,7 +22,7 @@
               "notes": "If the <code>browser.display.use_document_colors</code> user preference in <code>about:config</code> is set to <code>false</code>, background images will not be displayed."
             },
             "firefox_android": {
-              "version_added": true
+              "version_added": "4"
```
Comment (ddbeck) [inline css/properties/background-image.json]: I'm assuming this is based on the value for Firefox. If so we should pull down the note, or substantiate that the behavior is different for Android.

File: css/properties/mask.json [hunk 7]
```diff
@@ -27,11 +27,11 @@
               "version_added": true
             },
             "firefox": {
-              "version_added": true,
+              "version_added": "2",
               "notes": "From Firefox 10, the default color space when handling masks is sRGB. Previously, the default and only supported color space was linear RGB. This changes the appearance of mask effects, but brings Firefox into compliance with the second edition of the SVG 1.1 specification."
             },
             "firefox_android": {
-              "version_added": true
+              "version_added": "4"
```
Comment (ddbeck) [inline css/properties/mask.json]: I'm assuming this is based on the value for Firefox. If so we should pull down the note, or substantiate that the behavior is different for Android.

**Comment by vinyldarkscratch** [comment]
Oh whoops!  That script isn’t meant to be included in this PR, haha — it was supposed to be just a local copy, but looks like I committed it rather than stashing it...  😅 

As far as “48a459”, I’m not sure what exactly that means, would you be willing to explain?

File: css/properties/background-image.json (hunk 6 above)
Comment (vinyldarkscratch) [reply css/properties/background-image.json]: The assumption is correct.  Sure thing, pulling the notes down now!

**Comment by ddbeck** [comment]
> As far as “48a459”, I’m not sure what exactly that means, would you be willing to explain?

Sorry, I just flubbed copying the hash of the commit that added the script (GitHub will link to the specific commit like you would reference an issue, assuming you paste the right text, which I did not). Since you've removed the script from the PR, it doesn't matter. 😄

**Comment by vinyldarkscratch** [comment]
Ah, I see now, thanks for clarification!  ~~It’s the commit that was never meant to be~~

**Comment by ExE-Boss** [comment]
It’s commit https://github.com/mdn/browser-compat-data/commit/48a459ffc11398b124fedffd6868c269f27cd989 for those interested.

**Pull request merged**